src/mozanalysis/frequentist_stats/linear_models.py

```diff
@@ -0,0 +1,412 @@
+        res[low_str] = low
+        res[high_str] = high
+    return res
+
+
+def summarize_univariate(
+    data: pd.Series, branches: pd.Series, branch_list: list[str], alphas: list[float]
+) -> dict[str, pd.Series]:
+    """Univariate inferences (point estimates and confidence intervals) for the
+    mean of each branch's data.
+
+    Parameters:
+    - data (pd.Series): the vector of observations
+    - branches (pd.Series): the vector of branch names. Expected to be the same length
+    as `data`
+    - branch_list (list[str]): the list of branches to perform inferences on
+    - alphas (list[float]): the desired confidence levels
+
+    Returns:
+    - out (dict[str, pd.Series]): a dictionary keyed by branch label with a pandas
+    series as value. The format of the series is described above in
+    `summarize_one_branch`
+    """
+    return {
+        b: summarize_one_branch(data.loc[(branches == b).values], alphas)
+        for b in branch_list
+    }
+
+
+def _make_formula(target: str, ref_branch: str, covariate: str | None = None) -> str:
+    """Makes a formula which defines the model to build. Includes terms for
+    treatment branch and, optionally, a main effect for a (generally pre-experiment)
+    covariate.
+
+    Ex: given `target` of 'Y' and `ref_branch` of control, the formula will be:
+    `Y ~ C(branch, Treatment(reference='control'))`. That is, we want to predict `Y`
+    with treatment branch (`branch`) treated as a categorical predictor with
+    reference level of `'control'`.
+
+    This builds the following linear model (assuming 2 other treatment branches, t1
+    and t2): `Y_i = \beta_0 + \beta_t1*I(branch == 't1') + \beta_t2*I(branch == 't2')`
+    where `I` is the indicator function.
+
+    Inferences on `\beta_t1` are inferences of the average treatment effect (ATE) of
+    branch t1. That is, the confidence interval for `\beta_t1` is the confidence
+    interval for the ATE of branch t1. Similarly for t2 and `\beta_t2`.
+
+    We can incorporate a single (generally pre-experiment) covariate. Adding a covariate
+    of `Ypre` to the above, we'll build the following formula:
+    `Y ~ C(branch, Treatment(reference='control')) + Ypre`
+    which will fit the following linear model:
+
+    `Y_i = \beta_0 + \beta_t1*I(branch == 't1') + \beta_t2*I(branch == 't2') + Ypre_i`
+
+    For now, we elect to not include branch by covariate interaction terms and instead
+    to perform inferences only on the population-level ATE.
+
+    Parameters:
+    - target (str): the variable of interest.
+    - ref_branch (str): the name of the reference branch
+    - covariate (Optional[str]): the name of a covariate to include in the model.
+
+    Returns:
+    - formula (str): the R-style formula, to be passed to statsmodels's formula API.
+
+    """
+    formula = f"{target} ~ C(branch, Treatment(reference='{ref_branch}'))"
+    if covariate is not None:
+        formula += f" + {covariate}"
```

Comment:
If there are forbidden characters for the string, maybe have a check that `target`, `ref_branch` and `covariate` don't contain those strings?  Seems kind of insane someone would put a tilde in a variable names but couldn't hurt 🤷

Reply:
Might make more sense to put that check in `summarize_joint`

Reply:
Added!